File: solutions/python/intervals/meeting-rooms-ii.py
# LeetCode 253 — Meeting Rooms II (Medium)
# Category: Intervals · Approach: Sweep Line
# Time: O(n log n) | Space: O(n)
# Source: https://leetcode.com/problems/meeting-rooms-ii/

def min_meeting_rooms(intervals: list[list[int]]) -> int:
    starts = sorted(x[0] for x in intervals)
    ends = sorted(x[1] for x in intervals)
    rooms = max_rooms = e = 0
    for s in starts:
        while e < len(ends) and ends[e] <= s:
            rooms -= 1
            e += 1
        rooms += 1
        max_rooms = max(max_rooms, rooms)
    return max_rooms

File: solutions/typescript/intervals/meeting-rooms-ii.ts
// LeetCode 253 — Meeting Rooms II (Medium)
// Category: Intervals · Approach: Sweep Line
// Time: O(n log n) | Space: O(n)
// Source: https://leetcode.com/problems/meeting-rooms-ii/

function minMeetingRooms(intervals: number[][]): number {
  const starts = intervals.map((x) => x[0]).sort((a, b) => a - b);
  const ends = intervals.map((x) => x[1]).sort((a, b) => a - b);
  let rooms = 0, maxRooms = 0, e = 0;
  for (let i = 0; i < starts.length; i++) {
    while (e < ends.length && ends[e] <= starts[i]) {
      rooms--;
      e++;
    }
    rooms++;
    maxRooms = Math.max(maxRooms, rooms);
  }
  return maxRooms;
}
